### DISCUSSION TITLE
Speedup online session join by 5 times

### DISCUSSION THREAD

**Discussion by maybegreat48** [YimMenu/YimMenu#143] [Research]
I have finally figured out a solution for the delayed joining problem. All you have to do is native hook `NETWORK::NETWORK_HAS_RECEIVED_HOST_BROADCAST_DATA` and make it always return true

```
{ 0x5D10B3795F3FC886, &RETURN_TRUE },
```
This speeds up TRANSITION_STATE_FM_TRANSITION_CREATE_PLAYER from 1-2 minutes to 5-15 seconds. Haven't tried playing missions with it yet but it seems to be stable enough for use

**Comment by Yimura** [comment]
Definitely interesting, I'll have to take a look at the scripts where the game relies upon these natives. I would assume they have good reason to return false, but then again you say it doesn't negatively affect the experience?

Is this just native hooking for "freemode" or for all scripts?

**Reply by maybegreat48** [reply]
Am hooking it for all scripts

**Reply by DaniTRDev** [reply]
This native, and the broadcast natives are used by the game to sync specific RAGE variables across clients. If you return true when the game is still receiving the data this may lead to undefined behaviour in scripts. As my personal opinion, I won't recommend messing with this because it may fuck up online activities.

**Reply by Tayipc38** [reply]
Wouldn't that process still run since the native is only there to check whether its done?

**Reply by DaniTRDev** [reply]
The sync process would continue but if for some reason the game needs to get the sync data before it has finished the script will probably crash. This native is used to tell the game if it is ready to start "co-ops" activities and  players progression during this activities. For example, it is used in casino script.

**Reply by maybegreat48** [reply]
The process still runs yes, during initial connect with a script the array manager will have to download all the player/host broadcast data from all the players in the lobby/the script host. After everything is downloaded, it will sync only dirty elements to optimize buffer space. The broadcast is automatic and begins after a handshake with the script host

**Reply by DaniTRDev** [reply]
This native also allows to sync some globals between players. So it ""removes"" the need of being constantly syncing data.

**Reply by Tayipc38** [reply]
> The sync process would continue but if for some reason the game needs to get the sync data before it has finished the script will probably crash. This native is used to tell the game if it is ready to start "co-ops" activities and players progression during this activities. For example, it is used in casino script.

What if we hook it only when main_transition is running?

**Reply by maybegreat48** [reply]
The native has no side effects as far as I can see. The function that delays the script launch is run only when the script starts and not for every frame

**Reply by maybegreat48** [reply]
> > The sync process would continue but if for some reason the game needs to get the sync data before it has finished the script will probably crash. This native is used to tell the game if it is ready to start "co-ops" activities and players progression during this activities. For example, it is used in casino script.
> 
> What if we hook it only when main_transition is running?

That will certainly be a very good solution if this starts to cause any issues

**Reply by DaniTRDev** [reply]
That can work to avoid problems when joining. But I think that if you have set your respawn point at Casino you'll be telling the game that it is ready for everything in the casino script and that is not good (I didn't test this or anything like that, I'm just stating this using my sense of how this group of natives work).

**Reply by Yimura** [reply]
Sorry to revive this post again but in the meantime have you experienced any drawbacks from using this @maybegreat48 ?

**Reply by maybegreat48** [reply]
No. Online load is very stable. Note that however I have not tested exceptional cases like joining a friend in a mission from singleplayer for example

**Comment by Tayipc38** [comment]
Ive hooked the native in question in the script freemode and my results are as following:

The timing has been done from the press of the button to the end of the swoop down.
The swoop down has been known to be obsolete, in case this is also eliminated, 3-5 seconds can also be subtracted from each timing.

While testing I noticed a small pattern which might disrupt the validity of the timings. Since the timings have been done consecutively, an improvement is rather noticeable. This might be due to the fact that the game has loaded certain assets which are needed, thus not wasting time on any respective requirements of those same assets. To circumvent this,  I've rebooted the game before testing the Hook.

Syntax: Minutes:Seconds:Milliseconds
Conditions on hook: Always return true.

_Single player to online (Play GTA Online -> Go);_
**No Hook**
1. 1:02:71.
2. 0:53:74
3. 0:50:91

**With Hook**
1. 0:56.72
2. 0:56:51
3. 0:54:30

_Online find new session;_
**No Hook**
1. 0:41:77
2. 0:38:31
5. 0:33:23

**With Hook**
1. 0:37:21
2. 0:29:28
3. 0:34:70


**Verdict**:
Apart from the fact the game loads sessions faster the more sessions it has loaded, there does seem to be somewhat of an improvement. The Hook fires just before the transition "Freemode Transition Create Player" and skips a transition phase right before it. When used from single player to online, the improvement is less noticeable. In case of simply joining a new session when already online, there is a more significant improvement. To note, I've tested more after these initial ones and can confirm that sometimes the timing is under 25 seconds. The timing is really good when the joining method is substituted with a makeshift third party one involving script globals. 

Since these test have been done quickly and in a simple manner, I can't say whether hooking the native generates any byproducts.

**Reply by Yimura** [reply]
Any information of how many players were in the session at the time? I've only ever had this transition state `TRANSITION_STATE_FM_TRANSITION_CREATE_PLAYER` take longer with the more players there are in a session.

**Reply by Tayipc38** [reply]
Quite varying, but haven't paid much attention to it. Tests have been done on EU servers at a time where servers are mostly full.

**Reply by maybegreat48** [reply]
Yes this makes no difference in solo sessions but makes a huge difference in populated lobbies, this basically makes the host broadcast sync asynchronous instead of waiting for like a minute to receive it. You can also force script host of freemode during create player and that works as well but I assume it breaks the lobby as the broadcast data will be in an inconsistent state

**Comment by maybegreat48** [comment]
Some more optimizations include calling DLC::ON_ENTER_MP when the menu is loaded (in my case I use dll hijacking so as soon as main_persistent is started) and then suppressing all calls to DLC::ON_ENTER_SP and DLC::ON_ENTER_MP saves 5 seconds by removing the freeze in the clouds, You can also stop the player switch in TRANSITION_STATE_FM_SWOOP_DOWN, but I haven't really tried that yet

**Comment by maybegreat48** [comment]
This will make the session load superfast (combined with the native hook), but breaks RID join
```c++
// run every tick
if (*script_global::script_global(1574988).as<TransitionState*>() == TransitionState::TRANSITION_STATE_IS_FM_AND_TRANSITION_READY)
{
	STREAMING::STOP_PLAYER_SWITCH();
	*script_global::script_global(1574988).as<TransitionState*>() = TransitionState::TRANSITION_STATE_FM_FINAL_SETUP_PLAYER;
}

**Reply by xM4ddy** [reply]
> This will make the session load superfast (combined with the native hook), but breaks RID join

For you joining others via RID or others trying to RID join you?

**Reply by maybegreat48** [reply]
You joining others through friends menu or a RID joiner. Also forgot to mention that you will spawn with no weapons so you have to add some weapons to your ped

**Reply by Yimura** [reply]
Oh so it skips part of getting the cloud data, would be dope if we could asynchronously trigger it instead.

Also does this really only break RID join or any session join (with players)?

**Reply by maybegreat48** [reply]
Normal joining (through matchmaking etc.) is fine. Joining through pause menu (targeted join) is broken. I reckon it can be fixed though

**Reply by ZeddMOCO** [reply]
Thank you.

**Comment by Delfo19** [comment]
Hi, can you tell me how to add this function to the menu?

**Reply by Yimura** [reply]
It should already be implemented in YimMenu actually.

**Reply by Delfo19** [reply]
ok, i will see

**Reply by Delfo19** [reply]
is there an option to enable it or is enable by default?

**Reply by Yimura** [reply]
Always enabled as there are no known downsides.
